### PR TITLE
Support precise garbage collection with the Julia GC

### PR DESCRIPTION
**Pull request by fingolfin**
Under the Julia GC, GAP objects held in C locals are currently kept alive by conservatively scanning the C stack. This PR makes the kernel describe those roots precisely instead, the way Julia's own runtime does, so that the stack scanner is no longer needed. It is opt-in: configuring with `--with-gc=julia --enable-precise-gc` selects precise mode, and the define reaches kernel extensions through `sysinfo.gap`. Every other configuration is untouched: under GASMAN, Boehm, and Julia with stack scanning, all the new macros and annotations compile to nothing.

## What changes

**Root frames.** A C local that holds a GAP object across an allocation is registered with the collector: `GAP_GC_PUSH1(&x)` .. `GAP_GC_PUSH9(...)` push a frame of addresses, `GAP_GC_PUSHARGS` a frame of values for C arrays, and `GAP_GC_POP()` pops it (`src/precise_gc.h`). The macros wrap Julia's `JL_GC_PUSH*`.

Note that in order to support GAP immediate objects (such as produced by `INTOBJ_INT`), aka tagged pointers, a patch for the Julia GC is needed: https://github.com/JuliaLang/julia/pull/62889 makes it skip them in both kinds of frames, and is *required* for this PR, which uses both. I hope to get this backported.

**Annotations for the analyzer.** `GAP_GC_NOTSAFEPOINT`, `GAP_GC_CANSAFEPOINT`, `GAP_GC_ROOTED_BY_ARG(n)`, `GAP_GC_GLOBALLY_ROOTED` and friends let Julia's GC static analyzer (a clang plugin that Julia uses on its own runtime in order to find GC unsafe code) check the kernel: every function that can collect is marked, every value held across such a call must be rooted. Two mechanical commits carry most of the annotations; the rest of the series is the rooting the analyzer then demanded.

To fully make use of the analyzer part requires a second patch for Julia, see <https://github.com/JuliaLang/julia/pull/62928>

**Error unwinding.** A GAP error `longjmp`s past the frames the aborted call chain pushed. #6556 (merged) added hooks to the TRY/CATCH machinery to save and restore the collector's state; this PR uses them.

**Memory-checking mode** (`--enable-memory-checking`, `--enableMemCheck`), which so far was a GASMAN feature, now works under the Julia GC: it collects every nth allocation (`GAP_MEMCHECK_PERIOD`), mixes young and full collections (`GAP_MEMCHECK_FULL_EVERY`, the young ones are what exposes a missing write barrier), and aborts at the first reference the collector cannot vouch for, printing the parent bag, the slot, the GAP call stack and the registered frames. It also records every push and pop and reports frames an unwind left behind. This is what found the issues below; the recipe is in `dev/julia-gc-handoff.md`.

**CI.** A new workflow builds GAP in precise mode against a Julia with #62889 (juliaup channel `pr62889`) and runs the test suite optimized, in debug mode and under a memory-checking sample. It also runs the GC analyzer, the first-declaration check and the safepoint check, with the analyzer plugin built from https://github.com/JuliaLang/julia/pull/62928.

**Kernel issues.** Precise rooting has no tolerance for objects held only in C memory, and the checks found a number of such places. The ones that also matter under GASMAN were split out and are merged (#6551, #6552, #6553, #6554), as is the repair of the `DISABLE_STACK_SCAN` build (#6566). The remaining ones are Julia-only in effect but real: list elements held across a resize that drops them (`Remove`), the temporaries of the sorting kernels, the type array in method dispatch, the objects behind open files, `NewPlistFromArgs`' argument array, and `NewBag` itself, which held the fresh masterpointer only in a C local while it allocated the body.

## What is missing resp. to be done?

- more stress testing to uncover and fix more issues (or increase our confidence there are none... ho hum)
- need https://github.com/JuliaLang/julia/pull/62889 merged into Julia (and ideally backported)
- https://github.com/JuliaLang/julia/pull/62928 merged (not strictly necessary but greatly preferable)
- Packages with kernel extensions are not yet converted, so precise mode is not usable with them yet. JuliaInterface is the first: oscar-system/GAP.jl#1438 (draft) roots and annotates it, and a small header of no-op fallbacks keeps it building against older GAP kernels, so packages can merge such patches before this PR is merged. It builds on oscar-system/GAP.jl#1437, which stops GAP errors from `longjmp`ing across Julia frames.
- integrate with ongoing Windows / mingw work (possibly in a follow-up)
- incorporate reviewer feedback :-)

AI disclosure: prepared with Claude Code (Fable 5.1, Opus 5), which wrote code, tests, commit messages and drafts of this description.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by lgoettgens** [comment]
Same as https://github.com/oscar-system/GAP.jl/pull/1426#issuecomment-5546485489

**Comment by lgoettgens** [comment]
> * need [clangsa: Mark GC-tracked types with an attribute JuliaLang/julia#62928](https://github.com/JuliaLang/julia/pull/62928) merged into Julia (and ideally backported)
> * [clangsa: Mark GC-tracked types with an attribute JuliaLang/julia#62928](https://github.com/JuliaLang/julia/pull/62928) merged (not strictly necessary but greatly preferable)

I am not sure I follow the AI 😂
It states that this PR needs to be merged and ideally be backported, but merging is not strictly necessary

**Comment by fingolfin** [comment]
> I am not sure I follow the AI 😂

That wasn't the AI, that was me making a copy & paste mistake. I do not post un-edited AI generated PR descriptions, this one was already heavily edited, but humans make different mistakes compared to AI :-)

**Comment by ChrisJefferson** [comment]
I realize GAP-Julia integration is popular but I don't love this, it really does tie GAP into the Julia ecosystem -- After this is merged are any kernel developers going to be expected to understand how Julia's garbage collector annotations work and not break this?

I'm not saying I'd reject it, but I want to consider if there are different ways to do this, because we definitely wouldn't want to do it twice!

**Comment by ChrisJefferson** [comment]
One thing to consider if we are going to all this work -- should GASMAN use this too? I think the infrastructure isn't complicated (just making a big static array to push/pop the GC annotations into) and would make scanning easier. In particular it would fix the occasional hacks we have to do to work around weird OS/CPU issues, and would make wasm work better.

I wouldn't add this just for GASMAN, but if we are doing it anyway, might be worth considering if we should just use it everywhere (eventually).

File: src/error.c [hunk 1]
```diff
@@ -189,9 +190,9 @@ static Obj FuncPRINT_CURRENT_STATEMENT(Obj self,
                                        Obj context,
                                        Obj activeContext,
                                        Obj level,
-                                       Obj prefixWidth)
+                                       Obj prefixWidth) GAP_GC_CANSAFEPOINT
 {
-    volatile Obj location = Fail;
```
Comment (ChrisJefferson) [inline src/error.c:194]: I happened to spot this on the way by, we can't remove volatiles, because they are needed for longjmp/setjmp to work. There are a few removed.

**Comment by fingolfin** [comment]
> it really does tie GAP into the Julia ecosystem

It doesn't, and it is worth being precise about why. Everything this PR adds is GAP's own spelling: `GAP_GC_PUSH*/GAP_GC_POP` and the `GAP_GC_*` annotations. Under GASMAN and Boehm they compile to nothing; GAP builds and runs standalone exactly as before. Under the Julia GC the push/pop macros expand to Julia's frame macros, and the annotations expand to nothing there too. They only feed a static analyzer, run by a CI job this PR adds.

> are any kernel developers going to be expected to understand how Julia's
> garbage collector annotations work and not break this?

No. If the analyzer job fails on someone's kernel PR, I will fix it. That is the whole obligation, and it is mine. The annotations are documented in `src/precise_gc.h` for anyone who wants to understand them, but nobody has to.

> I want to consider if there are different ways to do this

As you know, for C there are three ways to find the roots on the C stack:
1. scan it conservatively (what we do now; the source of the OS/CPU hacks you mention, no wasm, and it depends on Julia continuing to support it),
2. compiler-emitted stack maps (not available for C), or
3. register the roots explicitly, which is this PR.
 
The last is the only precise option, and the part that costs effort is placing the ~650 push/pop pairs and the annotations. Strictly speaking the analyzer annotations are not necessary for this (they are compiled away after all); but they plus the clang static analyzer plugin written by the Julia team is what made this entire PR feasible: it allows us to determine where a root is needed, rather than guessing everywhere. So I'd say they are crucial after all, especially for maintaining the rest going forward (so they are not optional).

> should GASMAN use this too?

Yes, that is a direction I would be interested in working on. A GASMAN variant needs a redefinition of about a dozen macros; the push sites and annotations are shared.

Merging this PR is only the first step. Package kernel extensions need the same treatment, and precise mode stays opt-in until they have it, so nobody is affected who does not choose to be. Rolling it out will find more places that hold objects only in C memory; four of those are already fixed in master (#6551 to #6554), and I expect the rest to be of the same kind: pre-existing bugs the memory-checking mode now makes visible.

If it ever turns out to be a dead end, undoing it is cheap: everything here compiles away outside precise mode, and the annotations and push/pop statements are self-contained lines a script can strip.

File: src/error.c (hunk 1 above)
Comment (fingolfin) [reply src/error.c:194]: indeed, I already reverted some of those, but clearly missed some. should be fixed now, thanks for paying close attention
